pkg/controller/master/storagenetwork/storage_network.go

```diff
@@ -0,0 +1,310 @@
+	StorageNetworkAnnotation        = "storage-network.settings.harvesterhci.io"
+	ReplicaStorageNetworkAnnotation = StorageNetworkAnnotation + "/replica"
+	PausedStorageNetworkAnnotation  = StorageNetworkAnnotation + "/paused"
+
+	longhornStorageNetworkName = "storage-network"
+
+	// Rancher monitoring
+	CattleMonitoringSystemNamespace = "cattle-monitoring-system"
+	RancherMonitoringPrometheus     = "rancher-monitoring-prometheus"
+	FleetLocalNamespace             = "fleet-local"
+	RancherMonitoring               = "rancher-monitoring"
+	RancherMonitoringGrafana        = "rancher-monitoring-grafana"
+)
+
+type Handler struct {
+	ctx                  context.Context
+	longhornSettings     ctllonghornv1.SettingClient
+	longhornSettingCache ctllonghornv1.SettingCache
+	longhornVolumes      ctllonghornv1.VolumeClient
+	longhornVolumeCache  ctllonghornv1.VolumeCache
+	prometheus           ctlmonitoringv1.PrometheusClient
+	prometheusCache      ctlmonitoringv1.PrometheusCache
+	deployments          v1.DeploymentClient
+	deploymentCache      v1.DeploymentCache
+	managedCharts        ctlmgmtv3.ManagedChartClient
+	managedChartCache    ctlmgmtv3.ManagedChartCache
+	settings             ctlharvesterv1.SettingClient
+	settingsController   ctlharvesterv1.SettingController
+}
+
+// register the setting controller and reconsile longhorn setting when storage network changed
+func Register(ctx context.Context, management *config.Management, opts config.Options) error {
+	settings := management.HarvesterFactory.Harvesterhci().V1beta1().Setting()
+	longhornSettings := management.LonghornFactory.Longhorn().V1beta1().Setting()
+	longhornVolumes := management.LonghornFactory.Longhorn().V1beta1().Volume()
+	prometheus := management.MonitoringFactory.Monitoring().V1().Prometheus()
+	deployments := management.AppsFactory.Apps().V1().Deployment()
+	managedCharts := management.RancherManagementFactory.Management().V3().ManagedChart()
+
+	controller := &Handler{
+		ctx:                  ctx,
+		longhornSettings:     longhornSettings,
+		longhornSettingCache: longhornSettings.Cache(),
+		longhornVolumes:      longhornVolumes,
+		longhornVolumeCache:  longhornVolumes.Cache(),
+		prometheus:           prometheus,
+		prometheusCache:      prometheus.Cache(),
+		settings:             settings,
+		settingsController:   settings,
+		deployments:          deployments,
+		deploymentCache:      deployments.Cache(),
+		managedCharts:        managedCharts,
+		managedChartCache:    managedCharts.Cache(),
+	}
+
+	settings.OnChange(ctx, ControllerName, controller.OnStorageNetworkChange)
+	return nil
+}
+
+// webhook needs check if VMs are off
+// webhook needs check if nad is existing
+func (h *Handler) OnStorageNetworkChange(key string, setting *harvesterv1.Setting) (*harvesterv1.Setting, error) {
+	if setting == nil || setting.DeletionTimestamp != nil || setting.Name != settings.StorageNetworkName {
+		return nil, nil
+	}
+
+	if setting.Value == h.getLonghornStorageNetwork() {
```

Comment:
getLonghornStorageNetwork converts error to "", is it fit for an valid setting.Value ?

Reply:
let me check about this.